### PR TITLE
Validate admin webhook limit query forms

### PR DESCRIPTION
**Pull request by SYZBZ**
## Summary

- Reject ambiguous/non-canonical admin webhook limit query values before FastAPI scalar coercion.
- Apply the existing repeated-query and canonical-integer guards to `/admin?webhook_limit=...` and `/api/v1/admin/webhook-events?limit=...`.
- Preserve valid `1` and `200` limit values and existing admin-token behavior.

## Evidence

- Confusion, missing step, stale example, or bug this addresses: source issue #817 shows the admin webhook limit controls accepting `01`, `+1`, `1.0`, repeated values, and control-padded values as successful requests.
- Bounty capacity and active attempts/open PRs checked: Bounty #761 is live with `mrwk:bounty`, `Reserved on MergeWork`, API status `open`, and 6 effective awards remaining. `gh pr list --repo ramimbo/mergework --state open --search 817` returned no matching open PRs.
- Intended files or surfaces: `app/admin_routes.py`, `app/bounty_api.py`, `tests/test_admin_routes.py`, `tests/test_bounty_api.py`.
- Expected PR size: small, focused admin/API query validation fix.
- Out of scope: no admin authentication changes, webhook processing changes, treasury/proposal execution, payout behavior, wallet behavior, ledger behavior, private data, bridge/exchange/cash-out behavior, or MRWK price/liquidity claims.

## Test Evidence

- [x] `.\.venv\Scripts\python -m pytest tests\test_admin_routes.py tests\test_bounty_api.py tests\test_security.py -q` -> 92 passed, 1 warning
- [x] `.\.venv\Scripts\python -m pytest -q` -> 720 passed, 1 warning
- [x] `.\.venv\Scripts\python -m ruff check app\admin_routes.py app\bounty_api.py tests\test_admin_routes.py tests\test_bounty_api.py`
- [x] `.\.venv\Scripts\python -m ruff format --check app\admin_routes.py app\bounty_api.py tests\test_admin_routes.py tests\test_bounty_api.py`
- [x] `.\.venv\Scripts\python -m mypy app` -> Success: no issues found in 39 source files
- [x] `.\.venv\Scripts\python scripts\docs_smoke.py` -> docs smoke ok
- [x] `git diff --check`

## MRWK

Bounty #761
Source issue: #817
/claim #761

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened validation for admin webhook limit query parameters (reject repeated parameters and non-canonical integer encodings such as leading zeros, plus sign, floats, and control characters); returns proper 400 errors while preserving valid canonical values.

* **Tests**
  * Added comprehensive tests covering rejection of invalid/duplicate limit values and acceptance of canonical limit values for both admin pages and API endpoints.

### CODE REVIEW

File: app/bounty_api.py [hunk 1]
```diff
@@ -244,11 +244,14 @@ def api_bounties_summary(
...
     @app.get("/api/v1/admin/webhook-events")
     def api_admin_webhook_events(
+        request: Request,
         status: str | None = Query(None),
         limit: Annotated[int, Query(ge=1, le=200)] = 50,
         admin_login: str = Depends(require_admin_token),
     ) -> list[dict[str, Any]]:
         del admin_login
+        reject_repeated_query_param(request, "limit")
+        reject_noncanonical_int_query_param(request, "limit")
```
[the remark this answers is not in the document]
Comment (SYZBZ) [reply app/bounty_api.py:255]: Thanks for checking this. I verified this path with the new focused tests before and after the implementation. On the current head, `limit=01`, `limit=%2B1`, `limit=1.0`, repeated `limit`, and `%C2%851` all return the intended 400 detail from the route-level guards; they do not fall through to 422 in this FastAPI/Pydantic setup. The focused run was `python -m pytest tests/test_bounty_api.py::test_admin_webhook_events_api_rejects_noncanonical_limit_values -q`, and the broader `tests/test_bounty_api.py tests/test_admin_routes.py tests/test_security.py` run also passed.

File: app/admin_routes.py [hunk 2]
```diff
@@ -52,6 +56,8 @@ def admin_page(
         webhook_limit: Annotated[int, Query(ge=1, le=max(ADMIN_WEBHOOK_LIMIT_OPTIONS))] = 25,
         proposal_id: Annotated[int | None, Query(ge=1)] = None,
     ) -> Any:
+        reject_repeated_query_param(request, "webhook_limit")
+        reject_noncanonical_int_query_param(request, "webhook_limit")
```
[the remark this answers is not in the document]
Comment (SYZBZ) [reply app/admin_routes.py:59]: Thanks. I checked this explicitly with the new admin-page tests. The invalid `webhook_limit` spellings hit the route-level guards and return the intended 400 responses on the current head, including the C1-control case, rather than a framework 422. The broader admin/security test run passed as well.

File: tests/test_bounty_api.py [hunk 3]
```diff
@@ -122,6 +122,42 @@ def test_admin_webhook_events_api_rejects_c1_control_status(monkeypatch):
...
+@pytest.mark.parametrize(
+    ("query", "detail"),
+    [
+        ("limit=01", "limit must be a canonical positive integer"),
+        ("limit=%2B1", "limit must be a canonical positive integer"),
+        ("limit=1.0", "limit must be a canonical positive integer"),
+        ("limit=bad&limit=1", "limit must be provided at most once"),
+        ("limit=%C2%851", "limit must not contain control characters"),
+    ],
+)
```
[the remark this answers is not in the document]
Comment (SYZBZ) [reply tests/test_bounty_api.py:147]: Fair point. I kept the matrices local for this small focused PR because the param names and expected messages differ and there are only two route surfaces. If more admin/query surfaces reuse this exact matrix, a shared helper would be a good follow-up; for this bounty fix I kept the test shape explicit and easy to scan.

**Review by jakerated-r — APPROVED** [review]
Reviewed current head `3cea016e26c9b5427f382590f971f4243989c8c9` as a non-author against current `origin/main` `12ca2f0510abce4f3a8b3d77ae8596f47fa5c74c`.

Files inspected: `app/admin_routes.py`, `app/bounty_api.py`, `tests/test_admin_routes.py`, and `tests/test_bounty_api.py`.

What I checked:
- the admin HTML route now rejects repeated and non-canonical `webhook_limit` values before passing the value into `admin_page_context`;
- the admin webhook-events API route now applies the same repeated/canonical integer guards to `limit`;
- both surfaces preserve valid edge values `1` and `200`;
- the new tests exercise repeated values, leading-zero, plus-sign, decimal, and C1-control cases for both routes.

Validation:
- `PYTHONPATH=. uv run --extra dev pytest tests/test_admin_routes.py tests/test_bounty_api.py tests/test_security.py -q` -> `92 passed, 1 warning`
- `PYTHONPATH=. uv run --extra dev pytest -q` -> `720 passed, 1 warning`
- `PYTHONPATH=. uv run --extra dev ruff check app/admin_routes.py app/bounty_api.py tests/test_admin_routes.py tests/test_bounty_api.py` -> passed
- `PYTHONPATH=. uv run --extra dev ruff format --check app/admin_routes.py app/bounty_api.py tests/test_admin_routes.py tests/test_bounty_api.py` -> passed
- `PYTHONPATH=. uv run python scripts/docs_smoke.py` -> `docs smoke ok`
- `PYTHONPATH=. uv run --extra dev mypy app` -> success
- `git diff --check origin/main...HEAD` -> clean
- `git merge-tree --write-tree origin/main HEAD` -> clean tree `bca29238b80e9b08e557879b79fd6609e8bf1793`

I also reviewed the Copilot comments. The two concern comments about framework coercion do not reproduce under the branch's own invalid-value tests, which pass with the expected 400 responses. The shared-test-matrix suggestion is optional cleanup and not a blocker for this small patch.

Hosted status note: GitHub quality/readiness/docs/image checks are successful. CodeRabbit was still pending at review time, so this approval relies on local validation plus the completed hosted quality check.

No private data, wallet material, payout execution, treasury mutation, admin credential changes, exchange, bridge, cash-out, or MRWK price behavior was used or changed.

**Comment by ramimbo** [comment]
Holding this for another useful current-head non-author human review before acceptance. This touches admin/API behavior, so I am keeping the 2-review bar. CI and CodeRabbit are green; no `mrwk:accepted` or payout label has been applied.

**Review by vondutchi — CHANGES_REQUESTED** [review]
Reviewed current head `3cea016e26c9b5427f382590f971f4243989c8c9` as a non-author against current `origin/main` `53d73feef2ad8a4e44571863c73375d7193c9019`.

The branch-local implementation remains focused and validates the intended webhook limit behavior, but this PR is no longer merge-ready against the current base.

Evidence checked:
- Inspected `app/admin_routes.py`, `app/bounty_api.py`, `tests/test_admin_routes.py`, and `tests/test_bounty_api.py`.
- The PR diff is still narrow: repeated/non-canonical `webhook_limit` validation on `/admin`, repeated/non-canonical `limit` validation on `/api/v1/admin/webhook-events`, and tests for invalid values plus valid `1`/`200` bounds.
- Branch-local validation passes: `pytest tests/test_admin_routes.py tests/test_bounty_api.py tests/test_security.py -q` -> `92 passed, 1 warning`.
- Targeted Ruff check passed for the four touched files.
- Targeted Ruff format check passed for the four touched files.
- `python scripts/docs_smoke.py` -> `docs smoke ok`.
- `git diff --check origin/main...HEAD` passed.
- Current-base mergeability is blocked: `git merge-tree --write-tree origin/main HEAD` exits non-zero with content conflicts in `app/admin_routes.py`, `app/bounty_api.py`, and `tests/test_admin_routes.py`. GitHub reports `mergeStateStatus=DIRTY` / conflicting.
- Hosted `Quality, readiness, docs, and image checks` and CodeRabbit are successful for the PR head, so the blocker is specifically current-main integration.

Requested follow-up: rebase or merge current `main`, preserve both the existing current-main query guards and this PRs `webhook_limit`/`limit` guards, resolve the three touched-file conflicts, then rerun the focused admin/bounty API tests, docs smoke, Ruff check/format, and diff check.

No private data, credentials, wallet material, payout execution, ledger mutation, treasury mutation, exchange, bridge, cash-out, or price behavior was used or changed.

**Comment by SYZBZ** [comment]
Current-main integration follow-up is complete on head `90de046264f82853e2e0ce6839f40a1be0a6ffc9`.

What changed:
- rebased the branch onto current `main`;
- resolved the conflicts in `app/admin_routes.py`, `app/bounty_api.py`, and `tests/test_admin_routes.py`;
- preserved the current-main guards for `webhook_status`, `proposal_id`, and API `status` while keeping this PR's `webhook_limit` / API `limit` guards.

Validation after the rebase:
- `.venv\Scripts\python -m pytest tests\test_admin_routes.py tests\test_bounty_api.py tests\test_security.py -q` -> `95 passed, 1 warning`
- `.venv\Scripts\ruff check app\admin_routes.py app\bounty_api.py tests\test_admin_routes.py tests\test_bounty_api.py` -> passed
- `.venv\Scripts\ruff format --check app\admin_routes.py app\bounty_api.py tests\test_admin_routes.py tests\test_bounty_api.py` -> passed
- `.venv\Scripts\python scripts\docs_smoke.py` -> `docs smoke ok`
- `git diff --check` -> clean
- `git merge-tree --write-tree origin/main HEAD` -> clean tree `48300da29f03d8e80e5207f2e6143e45ef7a8390`

Hosted checks on the new head:
- Quality, readiness, docs, and image checks -> pass
- CodeRabbit -> pass / review skipped

GitHub now reports `mergeable=MERGEABLE`; no private data, credentials, wallet material, payout execution, ledger mutation, treasury mutation, exchange, bridge, cash-out, or price behavior was used or changed.

**Comment by SYZBZ** [comment]
CodeRabbit follow-up is complete on head `da44a8663e2db328ebe58d99cae58aabfc674314`.

Change made:
- Added `test_admin_webhook_events_api_rejects_repeated_status` for `/api/v1/admin/webhook-events?status=paid&status=closed`, asserting HTTP 400 and `status must be provided at most once`.

Validation after this commit:
- `.venv\Scripts\python -m pytest tests\test_bounty_api.py tests\test_admin_routes.py tests\test_security.py -q` -> `96 passed, 1 warning`
- `.venv\Scripts\ruff check tests\test_bounty_api.py app\bounty_api.py app\admin_routes.py tests\test_admin_routes.py` -> passed
- `.venv\Scripts\ruff format --check tests\test_bounty_api.py app\bounty_api.py app\admin_routes.py tests\test_admin_routes.py` -> passed
- `.venv\Scripts\python scripts\docs_smoke.py` -> `docs smoke ok`
- `git diff --check origin/main...HEAD` -> clean
- `git merge-tree --write-tree origin/main HEAD` -> clean tree `b86a97291e0b28a58f7ff4eeb64b0f67b6041e7d`

Hosted checks on the new head:
- Quality, readiness, docs, and image checks -> pass
- CodeRabbit -> pass / review skipped

Scope note: the `proposal_id` guard/test mentioned in the generated warning is current-base behavior, not a new line in this PR's `origin/main...HEAD` diff. This PR's added lines remain focused on `webhook_limit`, API `limit`, and the repeated API `status` regression test requested by CodeRabbit.

**Review by yui-stingray — APPROVED** [review]
Reviewed current head `da44a8663e2db328ebe58d99cae58aabfc674314` as a non-author after the rebase and repeated-status follow-up.

What I checked:
- The current-base conflicts called out on the previous head are resolved; this branch now merges cleanly with current `origin/main`.
- `/admin` now rejects repeated and non-canonical `webhook_limit` values before rendering the admin page, while keeping valid `1` and `200` values working.
- `/api/v1/admin/webhook-events` now rejects repeated `status`, repeated `limit`, and non-canonical `limit` values before listing events, while keeping valid `1` and `200` values working.
- The new regression test covers the repeated `status` case on the admin webhook-events API.

Validation:
- `pytest tests/test_bounty_api.py tests/test_admin_routes.py tests/test_security.py -q --capture=no` -> `96 passed, 1 warning`
- `ruff check tests/test_bounty_api.py app/bounty_api.py app/admin_routes.py tests/test_admin_routes.py` -> passed
- `ruff format --check tests/test_bounty_api.py app/bounty_api.py app/admin_routes.py tests/test_admin_routes.py` -> passed
- `mypy app` -> success
- `python scripts/docs_smoke.py` -> `docs smoke ok`
- `git merge-tree --write-tree origin/main HEAD` -> clean
- `git diff --check origin/main...HEAD` -> clean

I do not see a remaining blocker on the current head. The only residual note is the existing Starlette/httpx TestClient deprecation warning from the test environment, which is unrelated to this PR's behavior.

**Comment by ramimbo** [comment]
Maintainer update: merged as useful non-payable work in this older-clean-PR pass. The referenced #761 bounty is filled/closed with no effective award capacity, so there is no live #761 payable path for this PR. No `mrwk:accepted` label was applied and no payment proposal was created.
